pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/logtags"
+)
+
+// traceValue represents the trace information for a single registration.
+type traceValue struct {
+	traceID tracingpb.TraceID
+	spanID  tracingpb.SpanID
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	baseCtx withContext
+	// appRespSent and storageAppendRespSent track whether range messages have
+	// already been logged to log at most once. This limits the log from growing
+	// too large at a small risk of missing some messages in the case of dropped
+	// messages or reproposals.
+	// TODO(baptist): If we find we are missing important logs, consider adding
+	// a mode to disable these checks and allowing all messages to be logged.
+	appRespSent, storageAppendRespSent atomic.Bool
+}
+
+// logf logs the message to the trace context and the base context. The base
+// context is populated on the leaseholder and is attached to the SQL trace.
+func (t *traceValue) logf(depth int, format string, args ...interface{}) {
+	log.InfofDepth(t.ctx, depth+1, format, args...)
+	if t.baseCtx != nil {
+		if ctx := t.baseCtx.Context(); ctx != nil {
+			log.VEventfDepth(ctx, depth+1, 3, format, args...)
+		}
+	}
+}
+
+// tracingID attempts to balance uniqueness with readability by only keeping the
+// lower 16 bits of the trace and span.
+func tracingID(traceID tracingpb.TraceID, spanID tracingpb.SpanID) string {
+	return fmt.Sprintf("%x.%x", uint16(traceID), uint16(spanID))
+}
+
+// RaftTracer is a utility to trace the lifetime of raft log entries. It may
+// include some unrelated entries, since it does not consider the term. It
+// traces at most one MsgAppResp and MsgStorageAppendResp per index which is the
+// first one that is past our index entry. This limitation means it may not
+// capture all the relevant messages particularly if the term changes.
+//
+// The library will log in two different ways once to the standard cockroach log
+// and once to the SQL trace on the leader.
+// TODO(baptist): Look at logging traces on followers as well and sending back
+// to the leader in the MsgAppResp. It would need to be best effort, but might
+// still be useful.
```

Comment:
I propose we do this in another PR. To minimize this PR, its overhead, and the concerns around the leaseholder's tracing context (things like `baseCtx` and `withContext`), I suggest we scope this PR only to log the events into the log.

Reply:
Agreed about not doing follower traces as part of this PR. As mentioned, I'd rather not remove the baseCtx piece as I think it adds the most value.